File: 30-days-of-code/php/day-12-inheritance.php
<?php

class person {
    protected $firstName, $lastName, $id;
    
    public function __construct($first_name, $last_name, $identification) {
        $this->firstName = $first_name;
        $this->lastName = $last_name;
        $this->id = $identification;
    }
    
    function printPerson() {
        print("Name: {$this->lastName}, {$this->firstName}\n");
        print("ID: {$this->id}\n");
    }
}
class Student extends person {
    private $testScores;

    // Begin Solution
    //---------------------------------------------------------------------------------------
    public function __construct($first_name, $last_name, $identification, $testScores) {
      person::__construct($first_name, $last_name, $identification);
      $this->testScores = $testScores;
    }

    function calculate(){
      $sum = 0;
      $size = sizeof($this->testScores);
      for($i = 0; $i < $size; $i++){
          $sum += $this->testScores[$i];
      }
      $average = (int) ($sum / $size);
      if( $average >= 90 && $average <= 100){
          return 'O';
      } else if($average >= 80 && $average <= 90) {
          return 'E';
      } else if($average >= 70 && $average <= 80) {
          return 'A';
      } else if($average >= 55 && $average <= 70) {
          return 'P';
      } else if($average >= 40 && $average <= 55) {
          return 'D';
      } else {
          return 'T';
      }
    }
    //---------------------------------------------------------------------------------------
    // End Solution
}

$file = fopen("php://stdin", "r");

$name_id = explode(' ', fgets($file));

$first_name = $name_id[0];
$last_name = $name_id[1];
$id = (int)$name_id[2];

fgets($file);

$scores = array_map(intval, explode(' ', fgets($file)));

$student = new Student($first_name, $last_name, $id, $scores);

$student->printPerson();

print("Grade: {$student->calculate()}");
?>

File: 30-days-of-code/cpp/day-12-inheritance.cpp
#include <vector>

using namespace std;

class Person {
	protected:
		string firstName;
		string lastName;
		int id;
	public:
		Person(string firstName, string lastName, int identification){
			this->firstName = firstName;
			this->lastName = lastName;
			this->id = identification;
		}
		void printPerson(){
			cout<< "Name: "<< lastName << ", "<< firstName <<"\nID: "<< id << "\n"; 
		}
	
};

class Student :  public Person {
	private:
		vector<int> testScores;  
	public:
        // Begin Solution
        //---------------------------------------------------------------------------------------
        Student(string firstName, string lastName, int identification, vector<int> testScores): Person( firstName, lastName, identification){
            this->testScores = testScores;
        }

        char calculate(){
            int sum = 0;
            
            for(int i = 0; i < this->testScores.size(); i++){
                sum += this->testScores[i];
            }
            int average = (int) sum / this->testScores.size();
            if( average >= 90 && average <= 100){
                return 'O';
            } else if(average >= 80 && average <= 90) {
                return 'E';
            } else if(average >= 70 && average <= 80) {
                return 'A';
            } else if(average >= 55 && average <= 70) {
                return 'P';
            } else if(average >= 40 && average <= 55) {
                return 'D';
            } else {
                return 'T';
            }
        };
        //---------------------------------------------------------------------------------------
        // End Solution
};

int main() {
	string firstName;
  	string lastName;
	int id;
  	int numScores;
	cin >> firstName >> lastName >> id >> numScores;
  	vector<int> scores;
  	for(int i = 0; i < numScores; i++){
	  	int tmpScore;
	  	cin >> tmpScore;
		scores.push_back(tmpScore);
	}
	Student* s = new Student(firstName, lastName, id, scores);
	s->printPerson();
	cout << "Grade: " << s->calculate() << "\n";
	return 0;
}